sdc11073/wsdiscovery.py

```diff
@@ -813,126 +828,132 @@ def removeSourceAddr(self, addr):
-            if not readable:
-                break
-            sock = readable[0]
+
+    def _recv_messages(self):
+        """For performance reasons this thread only writes to a queue, no parsing etc."""
+        for key, events in self._full_selector.select(timeout=0.1):
+            sock = key.fileobj
             try:
                 data, addr = sock.recvfrom(BUFFER_SIZE)
             except socket.error as e:
-                print('socket read error', e)
+                self._logger.warning('socket read error %s', e)
                 time.sleep(0.01)
                 continue
             if self.isFromMySocket(addr):
                 continue
-            getCommunicationLogger().logDiscoveryMsgIn(addr[0], data)
+            self._read_queue.put((addr, data))
 
-            env = None
-            env = parseEnvelope(data, addr[0])
-            if env is None:  # fault or failed to parse
-                continue
 
-            mid = env.getMessageId()
-            if mid in self._knownMessageIds:
-                self._logger.debug('message Id %s already known. This is a duplicate receive, ignoring.', mid)
-                continue
+    def _run_q_read(self):
+        """Read from internal queue and process message"""
+        while not self._quitRecvEvent.is_set():
+            try:
+                incoming = self._read_queue.get(timeout=0.1)
+            except queue.Empty:
+                pass
             else:
-                self._knownMessageIds.appendleft(mid)
-                try:
-                    del self._knownMessageIds[-50]  # limit length of remembered message Ids
-                except IndexError:
-                    pass
-            iid = env.getInstanceId()
-            mid = env.getMessageId()
-            if iid:
-                mnum = env.getMessageNumber()
-                key = addr[0] + ":" + str(addr[1]) + ":" + str(iid)
-                if mid is not None and len(mid) > 0:
-                    key = key + ":" + mid
-                if not key in self._iidMap:
-                    self._iidMap[key] = iid
+                addr, data = incoming
+                getCommunicationLogger().logDiscoveryMsgIn(addr[0], data)
+
+                env = parseEnvelope(data, addr[0], self._logger)
+                if env is None:  # fault or failed to parse
+                    continue
+
+                mid = env.getMessageId()
+                if mid in self._knownMessageIds:
+                    self._logger.debug('message Id %s already known. This is a duplicate receive, ignoring.', mid)
+                    continue
                 else:
-                    tmnum = self._iidMap[key]
-                    if mnum > tmnum:
-                        self._iidMap[key] = mnum
+                    self._knownMessageIds.appendleft(mid)
+                    try:
+                        del self._knownMessageIds[-50]  # limit length of remembered message Ids

```

Comment:
I feel like I'm overlooking something. Is this actually the intent here? If a new message arrives, delete 50 known id's **every** time? That thing is probably empty all the time, actually detecting a duplicate is luck.

Reply:
This del statement deletes only one element at position -50. Because thic code is called every time a new item is added, this limits the length to 50.

Reply:
I actually misread. But once this deque contains 50 Elements, you'd evict the newly added Id every time, wouldn't you? Because it's prepended